rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/geo/GeofencingProcessor.java

```diff
@@ -0,0 +1,198 @@
+/**
+ * Copyright © 2016-2024 The Thingsboard Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.thingsboard.rule.engine.geo;
+
+
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import org.springframework.util.CollectionUtils;
+import org.thingsboard.common.util.JacksonUtil;
+import org.thingsboard.rule.engine.api.TbContext;
+import org.thingsboard.server.common.data.AttributeScope;
+import org.thingsboard.server.common.data.id.EntityId;
+import org.thingsboard.server.common.data.id.RuleNodeId;
+import org.thingsboard.server.common.data.kv.AttributeKvEntry;
+import org.thingsboard.server.common.data.kv.BaseAttributeKvEntry;
+import org.thingsboard.server.common.data.kv.JsonDataEntry;
+
+import java.util.*;
+import java.util.concurrent.ExecutionException;
+import java.util.stream.Collectors;
+
+import static java.util.Collections.emptyList;
+import static org.springframework.util.CollectionUtils.isEmpty;
+
+public class GeofencingProcessor {
+
+    private final TbGpsMultiGeofencingActionNodeConfiguration nodeConfiguration;
+    private final TbContext context;
+
+    private static final String GEOFENCE_STATES_KEY = "geofenceStates";
+    private static final String GEOFENCE_STATE_ATTRIBUTE_KEY_PREFIX = "geofenceState_%s";
+
+    public GeofencingProcessor(TbContext context, TbGpsMultiGeofencingActionNodeConfiguration nodeConfiguration) {
+        this.nodeConfiguration = nodeConfiguration;
+        this.context = context;
+    }
+
+    public GeofenceResponse process(Long ts, RuleNodeId ruleNodeId, EntityId entityId, List<EntityId> matchedGeofences) throws ExecutionException, InterruptedException {
+        List<EntityId> enteredGeofences = emptyList();
+        List<EntityId> leftGeofences = emptyList();
+        List<EntityId> insideGeofences = emptyList();
```

Comment:
Why do we need to initialize these lists as empty lists? Looks like you handle this in methods:

getLeftGeofences, getInsideGeofences, getEnteredGeofences

Reply:
Discussed personally